File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.java
import java.util.stream.*;
import java.util.*;
class Solution {
    public int solution(int cacheSize, String[] cities) {
        LinkedList<String> cach = new LinkedList<>();
        
        int execTime = 0;
        
        for(String str : cities){
            str = str.toLowerCase();
            boolean isExist = cach.contains(str);
            
            if(isExist){
                cach.remove(str);
                cach.addLast(str);
                
                execTime += 1;
                continue;
            }
            
            if(cacheSize == 0){
                execTime += 5;
                continue;
            }
            
            
            if(cach.size() >= cacheSize){
                cach.pollFirst();
            }
            
            cach.addLast(str);
            
            execTime += 5;
        }
        return execTime;
    }
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque

def solution(cacheSize, cities):
    cache = deque([])
    runtime = 0
    for c in cities:
        c = c.lower()
        if c not in cache:
            runtime += 5
            if cacheSize == 0:
                continue
                
            cache.append(c)
            if len(cache) > cacheSize:
                cache.popleft()
                
            
        else:
            runtime += 1
            cache.remove(c)
            cache.append(c)
            
    return runtime